### ISSUE TITLE
Invalid file information in SPDX documents

### ISSUE THREAD

**Issue by armintaenzertng** [tern-tools/tern#1240]
Note: This uses the new version of the SPDX generation introduced in #1233. The old version sports the same errors and a few more that have been already fixed in the new version. 

**Describe the bug**
SPDX outputs with file information have a number of validation issues:
- some files don't have a checksum (maybe this is only the case for empty files, so currently this resorts to the SHA1 of the empty string so that the `SpdxDocument` can at least be generated)
- some files have invalid SpdxIds like `SPDXRef-None-None` or `SPDXRef-v2"`
- some license references from `LicenseInfoInFile` are not present in the `ExtractedLicensingInfo` section

**To Reproduce**
I used `tern report -i golang:1.12-alpine -f spdxjson -sv 2.3 -o output.json` to produce the output and then ran `pyspdxtools -i output.json` on it (note that the validation takes a while due to large SPDX document).
I'm not sure whether `-x scancode` would also be required as I recall that the above command used to not produce any file information before. In case there are problems, I attached my `output.json` as [output.txt](https://github.com/tern-tools/tern/files/12104025/output.txt) (JSON format is not supported by GitHub, it seems).

**Error in terminal**
Here are the validation issues:
```
Unrecognized license reference: LicenseRef-21495e9. license_expression must only use IDs from the license list or extracted licensing info, but is: LicenseRef-21495e9
Unrecognized license reference: LicenseRef-1c734cf. license_expression must only use IDs from the license list or extracted licensing info, but is: LicenseRef-1c734cf
Unrecognized license reference: LicenseRef-1b79b75. license_expression must only use IDs from the license list or extracted licensing info, but is: LicenseRef-1b79b75
Unrecognized license reference: LicenseRef-fa9fd02. license_expression must only use IDs from the license list or extracted licensing info, but is: LicenseRef-fa9fd02
Unrecognized license reference: LicenseRef-39c3ee0. license_expression must only use IDs from the license list or extracted licensing info, but is: LicenseRef-39c3ee0
Unrecognized license reference: LicenseRef-21495e9. license_expression must only use IDs from the license list or extracted licensing info, but is: LicenseRef-21495e9
Unrecognized license reference: LicenseRef-4ccf56f. license_expression must only use IDs from the license list or extracted licensing info, but is: LicenseRef-4ccf56f
Unrecognized license reference: LicenseRef-45c771b. license_expression must only use IDs from the license list or extracted licensing info, but is: LicenseRef-45c771b
Unrecognized license reference: LicenseRef-ca2312b. license_expression must only use IDs from the license list or extracted licensing info, but is: LicenseRef-ca2312b
spdx_id must only contain letters, numbers, "." and "-" and must begin with "SPDXRef-", but is: SPDXRef-v2"-None
spdx_id must only contain letters, numbers, "." and "-" and must begin with "SPDXRef-", but is: SPDXRef-v2"-None
did not find the referenced spdx_id "SPDXRef-None-None" in the SPDX document
```

**Expected behavior**
Tern's generated SPDX documents with file information should be valid.

**Environment you are running Tern on**
Enter all that apply
- tern at https://github.com/tern-tools/tern/pull/1233/commits/047e1cbf12bb735e15c5efbb44138b4a2bb61e4a
- Ubuntu 22.0.4
- Python 3.10.6

**Comment by rnjudge** [comment]
Hmm, I don't see these errors in the current/old version of Tern's output when I run `tern report -i golang:1.12-alpine -f spdxjson -sv 2.3 -o output.json`:

```
(ternenv) [rose@fedora tern]$ tern report -i golang:1.12-alpine -f spdxjson -o output-golang.json
(ternenv) [rose@fedora ternenv]$ java -jar tools-java-1.1.7-jar-with-dependencies.jar Verify tern/output-golang.json 
WARNING: sun.reflect.Reflection.getCallerClass is not supported. This will impact performance.
This SPDX Document is valid.

```

**Comment by rnjudge** [comment]
I'll take a look at the output you pasted and the output I have, but this seems to be introduced with the latest changes.

**Comment by armintaenzertng** [comment]
The java-tools don't seem to pick up on all invalidities, please also check with `pyspdxtools -i output.json`.

Also, do you get the large (around 6MB) SPDX output?

**Comment by rnjudge** [comment]
@armintaenzertng I will try to generate a file with golang:1.12-alpine and see if I can re-create the errors you are seeing.

**Comment by rnjudge** [comment]
Running with the old changes, my SBOM with scancode metadata is 3.3MB. Running with the new changes, when I generate a scancode SBOM, It is 6.0MB. So it seems like there is extra metadata in there somewhere....

I do see one of the errors you are talking about with the old changes, though, even with the java tools:
`Analysis exception processing SPDX file: No SPDX element found for SPDX ID SPDXRef-None-None`

I'll take a look. I'm assuming its another issue related to Scancode's recent restructuring.

**Comment by armintaenzertng** [comment]
> It is true that a package may contain no files if files_analyzed is false but it may still contain other packages. This error is the majority of what I'm seeing.

Yes, I noticed this bug, too. This is fixed in the current 0.8.0rc3 release (the spdx-tools PR also includes that fixed release already, please update your local code to get the change).

**Comment by armintaenzertng** [comment]
> Running with the old changes, my SBOM with scancode metadata is 3.3MB. Running with the new changes, when I generate a scancode SBOM, It is 6.0MB. So it seems like there is extra metadata in there somewhere....

This is due to the `hasFiles` field being deprecated, see [here](https://github.com/spdx/spdx-spec/blob/844144b3785dbc6a35065eda3b9d36adda874540/schemas/spdx-schema.json#L349). All SPDXIDs from the `hasFiles` property are now represented as relationships, which have more lines than just the SPDXID.

**Comment by armintaenzertng** [comment]
@rnjudge: It turns out the java-tools pick up on the invalidities mentioned above, but only after the
`Analysis exception processing SPDX file: No SPDX element found for SPDX ID SPDXRef-None-None`
is fixed.
